cmd/ido2icingadb/main.go

```diff
@@ -0,0 +1,437 @@
+}
+
+// fillCache fills <f.Cache>/<history type>.sqlite3 (actually types[*].cacheFiller does).
+func fillCache() {
+	progress := mpb.New()
+	for _, ht := range types {
+		if ht.cacheFiller != nil {
+			ht.setupBar(progress)
+		}
+	}
+
+	types.forEach(func(ht *historyType) {
+		if ht.cacheFiller != nil {
+			ht.cacheFiller(ht)
+		}
+	})
+
+	progress.Wait()
+}
+
+// migrate does the actual migration.
+func migrate(c *Config, idb *icingadb.DB, envId []byte) {
+	endpointId := sha1.Sum([]byte(c.Icinga2.Endpoint))
+	idbTx := &sync.Mutex{}
+
+	progress := mpb.New()
+	for _, ht := range types {
+		ht.setupBar(progress)
+	}
+
+	types.forEach(func(ht *historyType) {
+		ht.migrate(c, idb, envId, endpointId, idbTx, ht)
+	})
+
+	progress.Wait()
+}
+
+// migrate does the actual migration for one history type.
+func migrateOneType[IdoRow any](
+	c *Config, idb *icingadb.DB, envId []byte, endpointId [sha1.Size]byte, idbTx *sync.Mutex, ht *historyType,
+	convertRows func(env string, envId, endpointId icingadbTypes.Binary,
+		selectCache func(dest interface{}, query string, args ...interface{}), ido *sqlx.Tx,
+		idoRows []IdoRow) (icingaDbInserts, icingaDbUpserts [][]any, checkpoint any),
+) {
+	var lastQuery string
+	var lastStmt *sqlx.Stmt
+
+	defer func() {
+		if lastStmt != nil {
+			_ = lastStmt.Close()
+		}
+	}()
+
+	selectCache := func(dest interface{}, query string, args ...interface{}) {
+		// Prepare new one, if old one doesn't fit anymore.
+		if query != lastQuery {
+			if lastStmt != nil {
+				_ = lastStmt.Close()
+			}
+
+			var err error
+
+			lastStmt, err = ht.cache.Preparex(query)
+			if err != nil {
+				log.With("backend", "cache", "query", query).
+					Fatalf("%+v", errors.Wrap(err, "can't prepare query"))
+			}
+
+			lastQuery = query
+		}
+
+		if err := lastStmt.Select(dest, args...); err != nil {
+			log.With("backend", "cache", "query", query, "args", args).
+				Fatalf("%+v", errors.Wrap(err, "can't perform query"))
+		}
+	}
+
+	var args map[string]interface{}
+
+	// For the case that the cache was older that the IDO,
+	// but ht.cacheFiller couldn't update it, limit (WHERE) our source data set.
+	if ht.cacheLimitQuery != "" {
+		var limit uint64
+		cacheGet(ht.cache, &limit, ht.cacheLimitQuery)
+		args = map[string]interface{}{"cache_limit": limit}
+	}
+
+	upsertProgress, _ := idb.BuildUpsertStmt(&IdoMigrationProgress{})
+
+	ht.bar.SetCurrent(ht.done)
+
+	// Stream IDO rows, ...
+	sliceIdoHistory(
+		ht.snapshot, ht.migrationQuery, args, ht.lastId,
+		func(idoRows []IdoRow) (checkpoint interface{}) {
+			// ... convert them, ...
+			inserts, upserts, lastIdoId := convertRows(
+				c.Icinga2.Env, envId, endpointId[:], selectCache, ht.snapshot, idoRows,
+			)
+
+			// ... and insert them:
+
+			if idb.DriverName() == driver.MySQL {
+				// Avoid MySQL error 1205 (Lock wait timeout exceeded; try restarting transaction)
+				// due to concurrent transactions upsert the same table (history).
+				idbTx.Lock()
+				defer idbTx.Unlock()
+			}
```

Comment:
Is this still necessary given that different queries are now executed which are more similar to regular history writes which don't have something like this either?

I left this running on my larger test database for a bit (actually writing data for 15 minutes now) with this lock commented out and didn't notice the error so far (I don't know how often it would even happen in the first place though).

Reply:
I'll set up my own long-running test and get back to you on Monday.